R/convert.R

```diff
@@ -0,0 +1,132 @@
+#'    the output Parquet file to. Should include the register name as the last
+#'    part of the path. E.g., `path/to/register_name/`.
+#'
+#' @returns Returns a character scalar with the path to the created Parquet
+#'    file(s) (`output_path`), so it can be used in a targets pipeline.
+#'
+#' @export
+#' @examples
+#' \dontrun{
+#' convert_to_parquet(
+#'   list_sas_registers(project_id="202020")),
+#'   "output/path/to/register_name")
+#' }
+convert_to_parquet <- function(path, output_path) {
+  # Initial checks.
+  fs::file_exists(path)
+  checkmate::assert_character(path)
+  checkmate::assert_character(output_path)
+  checkmate::assert_scalar(output_path)
+
+  # Read SAS files.
+  data <- read_sas_files(path) |>
+    # Add year column if possible.
+    add_year_col() |>
+    # Removes duplicates from the same source file.
+    dplyr::distinct()
+  # TODO: Might have to do some processing of the dates. If so, create a helper
+  # function for this.
+  # mutate(across(where(~inherits(.x, what = "date")), as.character))
+  # mutate(across(where(lubridate::is.Date), as.character))
+
+  # Write to Parquet, partitioned by year if year column exists.
+  if (length(path) > 1 & any(colnames(data) == "year")) {
+    fs::dir_create(fs::path(output_path))
+    arrow::write_dataset(
+      dataset = data,
+      path = output_path,
+      format = "parquet",
+      partitioning = "year"
+    )
+  } else {
+    output_path <- paste0(output_path, ".parquet")
+    arrow::write_parquet(
+      x = data,
+      sink = output_path,
+    )
+  }
+
+  # Success message.
+  cli::cli_alert_success(
+    "Successfully converted {.val {fs::path_file(path)}} to Parquet format and saved it in {.path {output_path}}."
+  )
+
+  output_path
+}
+
+#' Read SAS files
+#'
+#' This function reads one or more SAS files and add a `source_file` column
+#' indicating the file each row came from. This column is useful for tracking
+#' the origin of the row when combining multiple files. It also ensures that
+#' duplicate rows across different files are not removed during the
+#' de-duplication step in `convert_to_parquet()`.
+#'
+#' @param path A character vector with the absolute path to the SAS file(s).
+#'
+#' @returns A data frame with the contents of the SAS file(s) plus a
+#'    `source_file` column.
+#' @keywords internal
+read_sas_files <- function(path) {
+  purrr::map(path, \(file_path) {
+    haven::read_sas(file_path) |>
+      dplyr::mutate(source_file = as.character(file_path))
+  }) |>
+    purrr::reduce(dplyr::full_join)
+}
+
+#' Add year column to data if possible
+#'
+#' @param data A data frame with a `source_file` column to extract year from.
+#'
+#' @returns A data frame with a year column added if the year could be
+#'    determined.
+#' @keywords internal
+add_year_col <- function(data) {
+  year <- get_year_from_col(data$source_file)
+
+  if (all(!is.na(year) & year %in% 1969:2030)) {
```

Comment:
```suggestion
  next_year <- as.integer(format(Sys.Date(), "%Y")) + 1L
  if (all(!is.na(year) & year %in% 1969:next_year)) {
```

So that we don't hardcode a year into the code. 1969 is fine because the registers don't exist before that.